docker/initdb.d/setup.sql

```diff
@@ -0,0 +1,9 @@
+log_message('Setup: Activate CORS');
+update DB.DBA.HTTP_PATH set HP_OPTIONS = serialize(vector('browse_sheet', '', 'noinherit', 'yes', 'cors', '*', 'cors_restricted', 0))  where HP_LPATH = '/sparql';
+log_message('Setup: Declare namespaces');
+DB.DBA.XML_SET_NS_DECL ('ex', 'http://example.org/', 2);
```

Comment:
@TallTed , is there a way to load the namespaces from Turtle files instead?  This seems like a rather hokey way to do it.

Reply:
Any Turtle file may contain its own namespaces.

Those loaded via the (SQL) `DB.DBA.XML_SET_NS_DECL()` function are just the presets that are used when no declaration is found in a SPARQL query, as [shown for the DBpedia instance](https://dbpedia.org/sparql/?help=nsdecl).

Does that make this seem less hokey?

Reply:
Unfortunately no.  I understand that it is just setting default namespaces, but it would be much simpler for a user if a Turtle or SPARQL file could simply be provided to do so.  After all, that's likely where they'll come from anyway.

But this certainly isn't a show stopper.  I was just wondering if it could be simplified.

Reply:
I'm not sure if that would even be possible to implement because in a real-world situation you often load multiple files, which may have conflicting namespaces and then it is not defined, which one would take precedence.

Reply:
> may have conflicting namespaces

Pointer?
In practice - do prefix declarations (they aren't namespaces - that's XML!) get redeclared?

http://prefix.cc/

Reply:
Sorry, I'm not sure what you mean with pointer, but I try to clarify what I mean and how I understand and use the terms, feel free to correct me with the right terms in case I use the wrong ones:

* "prefix", for example "owl": the abbreviation of a namespace
* "namespace", for example "http://www.w3.org/2002/07/owl#": the unabbreviated form a prefix is mapped to

```DB.DBA.XML_SET_NS_DECL``` is used in setup.sql to defined default prefix to namespace mappings, in case a SPARQL query does not declare the prefix.

When we load N-Triples files in `setup.sql` then we have no other choice but to do it this way, because N-Triple files don't contain prefixes.
However when we load Turtle or RDF/XML then the question is if it is possible to load them from there.
My argument against that, besides that as far as I know Virtuoso does not support this, is that it it does not make sense to implement it because of the following case:

## Example

1. setup.sql contains ```ld_dir_all ('/rdf/', '*.ttl', 'http://example.org');```
2. the `rdf` directory contains a.ttl and b.ttl
3. a.ttl maps the empty prefix to "http://example.org/ontology/"
4. b.ttl maps the empty prefix to "http://example.org/resource/"

Now the namespace mapping for the empty prefix in Virtuoso would be undefined.

Reply:
I don't think that's a good reason for disallowing it.  The same problem already exists if the user does this:
```
DB.DBA.XML_SET_NS_DECL ('ex', 'http://example.org/ontology/', 2);
DB.DBA.XML_SET_NS_DECL ('ex', 'http://example.org/resource/', 2);
``
The tool could warn if a prefix is redefined differently, but should not warn if it is redefined to have the same value, because it's common to define the same prefixes (the same way) in different files.

Consider this a feature enhancement suggestion.  :)

Reply:
I would argue that at least in this case there can be a defined behavior because the order of statements is deterministic, while in the "*.ttl" example it is either undefined or at least less clear. But I have to pass this feature enhancement suggestion to @TallTed, because I'm not a Virtuoso developer :-)
I don't think there is a succinct way to implement this in the `setup.sql` script itself.

Reply:
So it's only the empty prefix?  (and `ex:` !)

For all others, what I see out there is strong consistency across datasets.

> one would take precedence.

A PREFIX in Turtle can redefine a prefix mid file. So when loading or concatenated, last prefix wins. Prefixes don't affect the data. Only the presentation.